src/shims/posix/linux/sync.rs

```diff
@@ -0,0 +1,88 @@
+    if !(4..=7).contains(&args.len()) {
+        throw_ub_format!("incorrect number of arguments for futex syscall: got {}, expected between 4 and 7 (inclusive)", args.len());
+    }
+
+    // The first three arguments (after the syscall number itself) are the same to all futex operations:
+    //     (int *addr, int op, int val).
+    // Although note that the first one is often passed as a different pointer type, e.g. `*const AtomicU32` or `*mut u32`.
+    let addr = this.deref_operand(args[1])?;
+    let op = this.read_scalar(args[2])?.to_i32()?;
+    let val = this.read_scalar(args[3])?.to_i32()?;
+
+    // The raw pointer value is used to identify the mutex.
+    // Not all mutex operations actually read from this address or even require this address to exist.
+    let futex_ptr = addr.ptr.assert_ptr();
+
+    let thread = this.get_active_thread();
+
+    let futex_private = this.eval_libc_i32("FUTEX_PRIVATE_FLAG")?;
+    let futex_wait = this.eval_libc_i32("FUTEX_WAIT")?;
+    let futex_wake = this.eval_libc_i32("FUTEX_WAKE")?;
+
+    // FUTEX_PRIVATE enables an optimization that stops it from working across processes.
+    // Miri doesn't support that anyway, so we ignore that flag.
+    match op & !futex_private {
+        // FUTEX_WAIT: (int *addr, int op = FUTEX_WAIT, int val, const timespec *timeout)
+        // Blocks the thread if *addr still equals val. Wakes up when FUTEX_WAKE is called on the same address,
+        // or *timeout expires. `timeout == null` for an infinite timeout.
+        op if op == futex_wait => {
+            if args.len() < 5 {
+                throw_ub_format!("incorrect number of arguments for FUTEX_WAIT syscall: got {}, expected at least 5", args.len());
+            }
+            let timeout = this.read_scalar(args[4])?.check_init()?;
+            if !this.is_null(timeout)? {
+                // FIXME: Implement timeouts. The condvar waiting code is probably a good example to start with.
+                // Note that a triggered timeout should have this syscall return with -1 and errno set to ETIMEOUT.
+                throw_ub_format!("miri does not support timeouts for futex operations");
+            }
+            // Check the pointer for alignment. Atomic operations are only available for fully aligned values.
+            this.memory.check_ptr_access(addr.ptr.into(), Size::from_bytes(4), Align::from_bytes(4).unwrap())?;
+            // Read an `i32` through the pointer, regardless of any wrapper types (e.g. `AtomicI32`).
+            let futex_val = this.read_scalar(addr.offset(Size::ZERO, MemPlaceMeta::None, this.machine.layouts.i32, this)?.into())?.to_i32()?;
+            if val == futex_val {
+                // The value still matches, so we block the trait make it wait for FUTEX_WAKE.
+                this.block_thread(thread);
+                this.futex_wait(futex_ptr, thread);
```

Comment:
For mutex operations we have `mutex_enqueue_and_block` as one call to make; would it make sense to also do that for futexes?

Reply:
I tried to follow the existing `condvar_wait`, as that one comes closest to what `futex_wait` does. (They both allow waiting with a timeout and signalling/waking waiting threads.) There's no `condvar_wait_and_block` (nor a `signal` function that also `unblock()`s).

I guess that might be because waiting and signalling might involve setting up and cancelling a timeout, while nothing in `src/sync.rs` deals with timeouts right now. But I'm not sure.

Reply:
Okay, makes sense to follow the condvar style.